### ISSUE TITLE
Corrupting non-html content

### ISSUE THREAD

**Issue by kuhnza** [intesso/connect-livereload#39]
When livereload middleware is installed I'm finding that non-html content such as generated PDFs are being corrupted. I believe this is because livereload is trying to inject the reload script snippet into the content.

I notice that the code currently checks the request accept header however wouldn't it be more appropriate to check the response content-type header before injecting the script instead?

**Comment by gsikorski** [comment]
Me too :( One day lost. I use dynamic content (tar), so your fix does not work.

**Comment by andineck** [comment]
@gsikorski which version of `connect-livereload` are you using? Could you provide some code in order to reproduce the error?

**Comment by gsikorski** [comment]
I am using latest version (0.5.3). My pseudo code is:
var tar = require('tar-fs');
var download = function (req, res) {
                res.set({
                        'Content-Description': 'File Transfer',
                        'Content-Type': 'application/x-tar',
                        'Content-Length': 512*152,
                        'Content-Disposition': 'attachment; filename="awesome.tar"'
                    })
                    .writeHead(200);

                tar.pack('/tmp/awesomeDir').pipe(res);
}

**Comment by andineck** [comment]
hmm, might be a problem when piping stuff. 
the current implementation is a bit of a hack . . .
a clean stream implementation would help i think.

**Comment by foxx** [comment]
Any idea on when this is going to be fixed?

**Comment by icompuiz** [comment]
In my connect-livereload configuration, I passed the paths or extensions I want to ignore 

```
require('connect-livereload')({ ignore: ['/api/files', 'download'] })
```

After several months of working around this issue, this single, non-hacky change fixed it.

**Comment by adityab** [comment]
I'm another person who spent about 4 hours trying to figure out why streaming out of GridFS was failing.
Of course, it was not. ;)

@icompuiz Thanks for your fix. I wasn't aware of the `ignore` option. Blacklisting `/api` fixed it for me.

**Comment by andineck** [comment]
The ignore option might be helpful, the problem is though that the paths are user specific, and every one would have to adjust it according to their project setup.

I think we should work toward a solution that is save for binary files as well, so that they don't get corrupted.

Can someone provide a PR with a test case that is corrupting binary files?

Also any stream oriented solution would be welcomed, I would be happy to merge it.

**Comment by x-cray** [comment]
Lost several hours tracking down the issue.

**Comment by eekboom** [comment]
So I, too, was bitten by this and spent quite some time on it.
For me it were Excel files that got corrupted.
It works fine in 0.4.1 and is broken in 0.5.0 as well as in 0.5.3.

Here are the request details. Note that the URL does not contain any indication of the file type, but the "Content-Disposition" header does.

```
General
Remote Address:127.0.0.1:9001
Request URL:http://localhost:9001/cxf-services/batch/job-executions/9943ab85-bfc3-4870-89e5-ced8d66863cf/details
Request Method:GET
Status Code:200 OK

Response Headers
HTTP/1.1 200 OK
server: Apache-Coyote/1.1
set-cookie: JSESSIONID=64D49FD361BE01234569789CCD111C069; Path=/acme-web/; HttpOnly
x-ua-compatible: IE=edge
content-disposition: attachment; filename=booking.xls
date: Tue, 06 Oct 2015 11:06:34 GMT
expires: -1
content-type: application/vnd.ms-excel
transfer-encoding: chunked
connection: close
```

**Comment by fonzy2013** [comment]
Hi! I should have looked at this before... 
Using the angular-fullstack there were quite a few modules to go through.

Here is a full example usinge the demo from [excelbuilderjs-node] (https://github.com/riyadhalnur/excelbuilderjs-node)  to build and serve an excel sheet from express.

Steps to reproduce:
1. npm install excelbuilderjs-node
2. download [testdata](https://raw.githubusercontent.com/riyadhalnur/excelbuilderjs-node/master/demo/testdata.json)
3. Run the following app.js:

```
'use strict';

var express = require('express');
var app = express();
var testData = require('./testdata.json');
var morgan = require('morgan');
var Excel = require('excelbuilderjs-node');


app.use(require('connect-livereload')({
    ignore: ['.xlsx']
  }));
app.use(morgan('dev'));


//This works but not without the ending... app.get('/getexcel.xlsx', function(req, res) {
app.get('/getexcel', function(req, res) {
  var demoWorkbook = Excel.createWorkbook();
  var stylesheet = demoWorkbook.getStyleSheet();

  var themeColor = stylesheet.createFormat({
    font: {
      bold: true,
      color: {theme: 3}
    }
  });

  var columns = [
    {id: 'id', name: 'ID', type: 'number', width: 10},
    {id: 'name', name:'Name', type: 'string', width: 50},
    {id: 'number', name: 'Number', type: 'number', width: 50},
    {id: 'address', name: 'Address', type: 'string', width: 50},
    {id: 'lat', name: 'Lat', type: 'number', width: 15}
  ];

  var worksheetData = [
    [
      {value: 'ID', metadata: {style: themeColor.id, type: 'string'}}, 
      {value: 'Name', metadata: {style: themeColor.id, type: 'string'}}, 
      {value: 'Number', metadata: {style: themeColor.id, type: 'string'}}, 
      {value: 'Address', metadata: {style: themeColor.id, type: 'string'}}, 
      {value: 'Lat', metadata: {style: themeColor.id, type: 'string'}}
    ]
  ].concat(testData);

  var demoList = demoWorkbook.createWorksheet({name: 'Demo List'});

  demoList.setRowInstructions({
    height: 20
  });

  demoList.setData(worksheetData);
  demoList.setColumns(columns);

  demoWorkbook.addWorksheet(demoList);

  var result = Excel.createFile(demoWorkbook);
  var data = new Buffer(result, 'base64');

  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64');
  res.setHeader('Content-Disposition', 'attachment; filename=' + 'demo.xlsx');
  res.end(data);
});




app.listen(3300);


console.log('Listening on port 3300');

```

**Comment by romain10009** [comment]
This issue drove me crazy for two days while working on zip streaming! Is there no way to block the injection of the script to html pages? I will try ignore "/api"

**Comment by PaulMougel** [comment]
@romain10009 You can use the `ignore` option, as described in this thread and [in the README](https://github.com/intesso/connect-livereload#use):
> `ignore` gives you the possibility to ignore certain files or url's from being handled by `connect-livereload`.

The parameter is a list of strings or Regex:

```js
  app.use(require('connect-livereload')({
    port: 35729,
    ignore: ['.js', '.svg']
  }));
```

The default is:

```js
ignore: [
    /\.js(\?.*)?$/, /\.css(\?.*)?$/, /\.svg(\?.*)?$/, /\.ico(\?.*)?$/, /\.woff(\?.*)?$/,
    /\.png(\?.*)?$/, /\.jpg(\?.*)?$/, /\.jpeg(\?.*)?$/, /\.gif(\?.*)?$/, /\.pdf(\?.*)?$/
  ]
```

If you want to ignore `/api`, you can add `/^\/api\/(.*)/` to this list.

**Comment by romain10009** [comment]
Yes sure, i can use this feature. I just dont understand what the point is
to try to inject the script tag inside a zip file. Would it not be better
to use a white-list allowing only html files to be injected instead of a
black-list to prevent such kind of trouble i ran into?

**Comment by treejanitor** [comment]
+1 Glad someone is documenting this one.  Killed a chunk of time for me as well.

**Comment by jasine** [comment]
+1

**Comment by diwu1989** [comment]
this is screwing us over on `.xlsx` files, would like to see microsoft document files also be in the default blacklist please

**Comment by fkarlsson** [comment]
Wow after HOURS of debugging chunked Transfer-Encoding I find this issue. I was trying to send a chunked response of a csv file. Adding `ignore: ['csv']` to my `Gruntfile` solved the issue for my development environment.

**Comment by fkarlsson** [comment]
Wow after HOURS of debugging chunked Transfer-Encoding I find this issue. I was trying to send a chunked response of a csv file. 

I tried adding `ignore: ['.csv']` to the middleware setup but that did not solve my problem. I have disabled the middleware in Express for now and use the LiveReload Chrome extension instead.

The problem is that `connect-livereload` hijacks the `response.send()` function and causes the response to buffer until the end. Anyone have any ideas? Maybe this should be a new issue.

**Comment by gagan-bansal** [comment]
I was having same issue. 

I was using `compression` middleware. By making it applicable only for 'production', I could resolve the issue.

```javascript
if (process.env.NODE_ENV === 'production') {
  app.use(compression({}));
}
```
